### PR TITLE
Support THD context parallelism in pretrain_gpt.py

### PR DESCRIPTION
**Pull request by stnie**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Use ContextParallelBatch to produce correctly padded per-sequence zigzag batches for packed THD inputs with context parallelism.

Add self-contained two-rank parity coverage for BSHD, packed THD, unequal sequence lengths, and MTP.

## Issue tracking

Linked issue: Related to issue #6156 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: pretrain_gpt.py [hunk 1]
```diff
@@ -179,35 +202,33 @@ def get_batch(data_iterator, vp_stage: Optional[int] = None):
...
+    return get_batches_on_this_cp_rank(
         batch,
+        boundary_layout=config.attention_cp_layout,
         is_hybrid_cp=is_hybrid_cp,
         cp_group=get_context_parallel_group(),
         hybrid_cp_group_func=get_hybrid_data_context_parallel_groups,
-        use_per_sequence_balancing=args.dataloader_inter_document_masking and not is_sft,
+        use_per_sequence_balancing=(
+            (batch.get('cu_seqlens') is not None and not is_hybrid_cp)
+            or (args.dataloader_inter_document_masking and not is_sft)
```
Comment (Connor-XY) [inline pretrain_gpt.py:220]: **[P1] Enable dynamic pipeline shapes for padded batches**

With inter-document masking, CP=2, PP≥2, and `cu_seqlens=[0,3,8]`, this path pads the batch to 12 tokens, producing 6 tokens per CP rank. Pipeline receivers still allocate `seq_length / CP = 4` tokens because `variable_seq_lengths` remains false without a packing scheduler. Please enable dynamic pipeline shape exchange for this path, or otherwise keep the communicated shapes consistent with the padded batch.

File: pretrain_gpt.py [hunk 2]
```diff
@@ -179,35 +202,33 @@ def get_batch(data_iterator, vp_stage: Optional[int] = None):
...
+        ),
+        sequence_parallel=config.sequence_parallel,
+        tp_group=mpu.get_tensor_model_parallel_group(),
+        tp_cp_group=(
+            mpu.get_tensor_and_context_parallel_group()
+            if config.sequence_parallel and config.tensor_model_parallel_size > 1
+            else None
+        ),
+        tokens_per_sample=args.seq_length,
```
Comment (Connor-XY) [inline pretrain_gpt.py:229]: **[P1] Preserve sample boundaries when padding changes token counts**

Passing the original `args.seq_length` as `tokens_per_sample` becomes unsafe after per-document padding. With CP=2, MBS=3, sequence length 8, and documents `[3,5]` per sample, the new path produces 18 local tokens. `_maybe_unflatten_for_moe()` then reshapes `[18,1,64]` into `[8,2,72]`, causing the router projection to fail. Please update the sample-boundary handling so padding cannot change the inferred hidden dimension.

File: pretrain_gpt.py [hunk 3]
```diff
@@ -316,67 +337,30 @@ def forward_step(data_iterator, model: GPTModel, return_schedule_plan: bool = Fa
...
+        )
+        batch = cp_batch.get_batch(attention_layout)
+        attention_mask = batch.get("attention_mask")
+        cu_seqlens = batch.get("cu_seqlens")
+        labels = batch.get("labels")
+        loss_mask = batch.get("loss_mask")
+        position_ids = batch.get("position_ids")
+        tokens = batch.get("tokens")
+        packed_seq_params = cp_batch.get_packed_seq_params(attention_layout)
+        padding_mask = batch.get("padding_mask")
```
Comment (Connor-XY) [inline pretrain_gpt.py:361]: **[P2] Propagate a mask for the inserted padding**

The padded zigzag helper zeros synthetic positions in `loss_mask`, but does not populate `padding_mask`, so this lookup returns `None`. MoE auxiliary losses and expert-bias updates therefore count the newly inserted padding as real tokens; masking the language-model loss does not exclude it from those calculations. Please construct and propagate a padding mask for this path, including on intermediate pipeline stages.

File: pretrain_gpt.py [hunk 4]
```diff
@@ -179,35 +202,33 @@ def get_batch(data_iterator, vp_stage: Optional[int] = None):
...
         is_hybrid_cp=is_hybrid_cp,
         cp_group=get_context_parallel_group(),
         hybrid_cp_group_func=get_hybrid_data_context_parallel_groups,
-        use_per_sequence_balancing=args.dataloader_inter_document_masking and not is_sft,
+        use_per_sequence_balancing=(
+            (batch.get('cu_seqlens') is not None and not is_hybrid_cp)
+            or (args.dataloader_inter_document_masking and not is_sft)
+        ),
+        sequence_parallel=config.sequence_parallel,
+        tp_group=mpu.get_tensor_model_parallel_group(),
```
Comment (Connor-XY) [inline pretrain_gpt.py:223]: **[P2] Preserve CP alignment when adding TP padding**

This activates a failing path for a supported odd TP size. With sequence parallelism, TP=3, CP=2, and `cu_seqlens=[0,1,24]`, `_build_thd_zigzag_metadata()` changes padded document lengths from `[4,24]` to `[4,26]` to satisfy TP divisibility. The last document is then no longer divisible by `2 * CP`, and the rank-order scatter raises an out-of-bounds error. Please make the padding satisfy both alignment requirements and add an odd-TP regression case.

File: tests/unit_tests/models/test_gpt_thd_cp_parity.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+def test_pretrain_gpt_thd_cp2_with_mtp_matches_thd_cp1():
+    """GPT MTP must consume the same zigzag THD view as the main attention block."""
+    if int(os.environ.get("WORLD_SIZE", "1")) < 2:
+        pytest.skip("THD context-parallel parity requires at least two ranks")
+
+    thd_cp1 = _run_pretrain_gpt_forward(cp_size=1, thd=True, mtp=True)
+    thd_cp2 = _run_pretrain_gpt_forward(cp_size=2, thd=True, mtp=True)
+
+    assert thd_cp1.shape == thd_cp2.shape
+    _assert_close(thd_cp2, thd_cp1)
```
Comment (Connor-XY) [inline tests/unit_tests/models/test_gpt_thd_cp_parity.py:309]: **[P2] Compare the auxiliary MTP loss or gradients**

These assertions compare only the main LM loss. The helper runs under `eval()` and `torch.no_grad()`, while `process_mtp_loss()` attaches the auxiliary loss through `MTPLossAutoScaler`, whose forward returns the main hidden states unchanged. Consequently, incorrect—even NaN—MTP auxiliary losses can leave this comparison passing. Please compare the auxiliary MTP losses or MTP gradients between CP1 and CP2 so the test validates MTP numerical parity.

**Comment by Connor-XY** [comment]
@stnie Following up on the September 16 review: the draft is still at `2878df8`, and all five inline findings remain unanswered. Please address the dynamic pipeline-shape mismatch and sample-boundary/reshape failures first, then the inserted-padding mask, combined TP/CP alignment, and MTP auxiliary-loss/gradient coverage. Please reply on those threads with the fixes and focused regression results when available so I can re-review.
